### ISSUE TITLE
Wrong working directory ... or: PSyclone install does not install psyclone.cfg in a local install

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#355]
When I do a ``pip install --user -e .``, it appears that no psyclone.cfg is installed in a local directory, while a system-wide install will copy this file to /usr/share/psyclone.

So far this worked for me because I have a custom config file in psyclone/tests/.psyclone, which is used in most tests. But I just stumbled across some tests do a chdir to a temporary directory, and then NOT a chdir back to the original directory. Which means that the next test that is run will have the previous test's temporary directory as working directory. If additionally the latter test reloads the config file (since that is what it is supposed to test), now psyclone cannot find my .../psyclone/tests/,psyclone config file anymore (since it only tests .psyclone/psyclone.cfg), so the tests fail with config file not found. Doing a system-wide install fixes the problem (since psyclone always checks /usr/share/psyclone).

As a first solution I have fixed up the tests in my current branch to do a chdir back to the original directory, but that only solves the problem partially: if a test in a temporary directory should fail an assert, the chdir back is not executed, and the next test may fail because it has an unexpected working directory again :(

Two solutions that I can see:
1) we fix the install process in case of a local installation so that the psyclone.cfg file is copied to $HOME/.local/share/psyclone. --> All tests would find this config file since it is always tested for in config find_file().
2) We add a fixture in conftest that is executed for ALL tests (or at least the ones that chdir to a directory) that will make sure the chdir back to the original directory at the end (or at the beginning of each test).

The first one seems to be the easier solution, but I don't understand why it is not installed in the first place :)

**Comment by hiker** [comment]
It looks like data files are not installed when using ``-e .`` I don't know enough of setup.py to fix this immediately, but one suggestion would be to also test for  __file__/../../config/psyclone.cfg?

**Comment by hiker** [comment]
Ah, you knew more than me :) 

I have a config file in test because psyclone could not find a config file (because I never did a pip install without -e ;)  ). It was the easy solution for me to get things working again months ago :)

So from my point of view, I would prefer to add \_\_file\_\_+"/../../config" to the search path. The big advantage of this is that a new version of the config file can be tested and committed from the same directory (psyclone/config) - for me it was already quite annoying that I had to modify the config file in tests, then copy it back to the version that is in git in order to commit it. Currently you would have to run a pip install --user after every change/fix to the config file. Of course, we as developer would have to be aware of that (i.e. once we do an install without -e, the installed config file would always be used, not the one in the psyclone tree).

Fixing the cwd for all tests ... a lot of work for not much gain. All tests work fine (if the right config file is found, which the suggestion above would actually do). Any test that needs to run in a special directory (i.e. in a temp directory) will chdir properly - so that whole problem only came up because my work around for not having a .../share/psyclone/psyclone.cfg (i.e. work around = creating a config file in the test directory) stopped working. And it is only one or two specific tests which delete the already loaded config file to test special config files.

**Comment by arporter** [comment]
OK, so could we have a fixture that alters the config search path to include __file__+"/../../config"? I like that idea as it insulates the test suite from whatever configuration file a user currently has installed - on reflection it's not right that a user's installation can impact the test suite.

**Comment by hiker** [comment]
I had a first go at this: the main problem is that if we just set the search path to be ``../../config`` (i.e. make sure all files in tests find the config file in the git repo) , then some of the tests in config_test fail (the ones that test the search path behaviour obviously ;) ), with no easy way to keep those tests, which I think would still be an important test.

After some digging I did the following in https://github.com/stfc/PSyclone/commit/dec7fe12f957450375c1504d015bab4255f447af
It appends ``../../config`` to the end(!) of the search path only if pytest is used ('pytest' in sys.modules 😄  ). This way normal use of PSyclone will not at all be affected, and all existing tests that modify the search path still work - it just happen that all tests of search paths only test if the relative ordering of the directories is as expected (e.g. cwd comes before share, ...). And all normal tests will use the config file from the PSyclone repo, which is great.

Only problem: if the developer should have a PSyclone version properly installed (i.e. without -e), there will be a PSyclone config file in .local/... somewhere, which will take precedence (and that _might_ cause test failures, depending on content). I've documented this behaviour and added a proper warning to the developers section.

Admittedly this solution is not ideal (because of the potential clash with an installed file). An alternative solution: in case of running pytest, we could add ``../../config`` to the beginning(!) of the search list. Nearly all search path tests then still work (since they only test relative ordering, ignoring the new directory). The only test failure: config_test, test_search_env. It tests that if an invalid path is specified in PSYCLONE_CONFIG, PSYCLONE_CONFIG is ignored and the first search path is used: ``cwd/.psyclone``. Instead of testing if ``cwd/.psyclone/psyclone.cfg`` is returned, we could test if ``../../config/psyclone.cfg`` is returned. That is mostly nice, though then we have no way of testing if ``cwd/.psyclone`` would be used in a stand-alone environment, which is the nice side-effect of this test.

Refactoring the config code by making the search path a static variable, that we could modify using monkeypatch might not work: the current implementation used getcwd() to assemble the path, which is important in some of the tests (since they change into a temporary directory for tests, so getcwd() changes from test to test). 

It's a bit of a hassle this bit of testing :(

**Comment by arporter** [comment]
Urgh, sorry it's proving so difficult. I think if we're going to this trouble then we should ensure that we don't unintentionally pick-up some previously installed config file. 

Could we use $PSYCLONE_CONFIG to ensure that `../../config/psyclone.cfg` is picked up in the majority of tests? I think we could set it via a fixture (yes, I like fixtures :-) ). That uses the existing mechanisms unchanged and can easily not be done to allow the tests that get upset to operate normally.

**Comment by hiker** [comment]
Thanks for the suggestion! Done in c5bf11df7af88fda704b14d9859799e7131f5f8d (as part of the #329 branch). It was actually a lot simpler than expected - I had added a setup and tear-down function using fixtures that make sure that PSYCLONE_CONFIG is always as expected, but it turns out all tests that test search path order and/or PSYCLONE_CONFIG use monkeypatch, so there was no need to set/unset PSYCLONE_CONFIG at all. All I needed was one **session** setup fixture to set this variable, which seems to be working fine ... for me :sob:  On travis it failed because the installation does NOT have the psyclone.cfg file in the expected location (root_dir/config/psyclone.cfg), for which I had added a new test:
```
    config_file = Config.get_repository_config_file()     ## This is `__file__`/../../config/psyclone.cfg
    assert "../../config/psyclone.cfg" in config_file
    assert os.path.isfile(config_file)   ## This fails in case of running the tests in an installation
```

All (other) tests worked since Configuration checks if PSYCLONE_CONFIG exists, and if not uses the normal search path (which means the installed psyclone.cfg file was found). Since it seems odd to set PSYCLONE_CONFIG to a file that does not exists, I am now only setting PSYCLONE_CONFIG if the repository config file exists, and I have therefore removed this last test.
